### DISCUSSION TITLE
🐛Indexing duplications

### DISCUSSION THREAD

**Discussion by 92amartins** [policratus/pupyl#63] [Ideas]
**Describe the bug**
I'm afraid in some cases pupyl is indexing the same image more than once.

**To Reproduce**
1. Collect few images (e.g. 10)
1. Run the example code from README.
1. See error

**Expected behavior**
On the screenshot below I expected `cars` to appear only once.

**Screenshots**
<img width="1436" alt="image" src="https://user-images.githubusercontent.com/4500163/102019229-03c7c080-3d51-11eb-9ebb-d799412bd06c.png">

**Desktop (please complete the following information):**
 - OS: macOS Big Sur
 - Browser safari 14.0.1
 - Python 3.8.6
 - Pupyl Version 0.10.4

**Additional context**
Folder Structure:
```
.
├── images
│   ├── batman.jpg
│   ├── captain.jpg
│   ├── cars.jpg
│   ├── flash.jpg
│   ├── green.jpg
│   ├── inside_out.jpg
│   ├── monsters.jpg
│   ├── nemo.jpg
│   ├── superman.jpg
│   └── toy.jpg
└── images.py
```

```
# image.py
from pupyl.search import PupylImageSearch
from pupyl.web import interface


SAMPLES = 'images'

SEARCH = PupylImageSearch()

SEARCH.index(SAMPLES)

interface.serve()
```

**Comment by policratus** [comment]
First of all, thank you @92amartins for reporting this. A question: did you run the method `PupylImageSearch.index()` twice in the same source (in this case, a directory)?

**Comment by 92amartins** [comment]
You're welcome.

Yes :(

In the first run, I had the images at the same directory as `images.py`. That caused `pupyl` to try to index the .py file raising `FileIsNotImage` exception. In spite of that, probably indexed a few files (alphabetically?) before raising the exception.

For the second run,  I moved all the images to a separate folder, then ran the script with no issues.

I think that's why it got duplicated.

Maybe `pupyl` could have a mechanism for preventing indexing duplicated images, but as I'm not sure if this is a main case or edge case, I'll leave it up to you.

Feel free to close the issue, as I clearly made a misuse.

**Comment by policratus** [comment]
> You're welcome.
> 
> Yes :(
> 
> In the first run, I had the images at the same directory as `images.py`. That caused `pupyl` to try to index the .py file raising `FileIsNotImage` exception. In spite of that, probably indexed a few files (alphabetically?) before raising the exception.
> 
> For the second run, I moved all the images to a separate folder, then ran the script with no issues.
> 
> I think that's why it got duplicated.
> 
> Maybe `pupyl` could have a mechanism for preventing indexing duplicated images, but as I'm not sure if this is a main case or edge case, I'll leave it up to you.
> 
> Feel free to close the issue, as I clearly made a misuse.

Well, we do have a bug here: `pupyl` should only try to index supported image files, not all files in a source. I'll open a issue to fix this bug.

The mechanism that you described seems pretty useful to avoid any duplications and it will be implemented as an optional parameter during an index process. This mechanism is nice but creates a little overhead during the index, so will be up to the user choosing it.

**Comment by policratus** [comment]
@92amartins, issues #66 and #67 created to fix or implement these bugs or ideias.

**Comment by policratus** [comment]
@92amartins, issue #67 resolved on pull request #68. Could you check if it works as intended now, when you have a chance?

**Reply by 92amartins** [reply]
Yeah, sure! Can you provide an alpha version or instructions for building from source?

**Reply by policratus** [reply]
Should be a good idea having another branch with alpha releases, isn't it? I'll organize this too.

So @92amartins, to test the changes as it is, follow the steps:

1. **Clone the branch**:
```shell
# This case, the branch is check-duplicates
git clone -b check-duplicates git@github.com:policratus/pupyl.git
```

2. **(Optional) Create a virtual environment**: Using [virtualenvwrapper](https://pypi.org/project/virtualenvwrapper/), issue:
```shell
mkvirtualenv pupyl
```

3. **Install dependencies**: In the cloned directory, install the package (and its dependencies) with
```shell
cd pupyl && python setup.py install
```

4. **(Optional) Compilation errors**: if you see C/C++ compilation errors like `error: command 'x86_64-linux-gnu-gcc' failed with exit status 1`, go to [TROUBLESHOOTING](https://github.com/policratus/pupyl/blob/master/TROUBLESHOOTING.md)

5. **(Optional) Create a directory with duplicated images**: These images can be used to test:

Original          |  Subtle difference     | Explicit difference
:-------------------------:|:-------------------------:|:-------------------------:
<img src="https://user-images.githubusercontent.com/827563/102917553-31ed8480-4464-11eb-88e5-19e7e832aadb.jpg" width="256">  | <img src="https://user-images.githubusercontent.com/827563/102917554-32861b00-4464-11eb-89d4-26888e28d235.jpg" width="256"> | <img src="https://user-images.githubusercontent.com/827563/102917555-331eb180-4464-11eb-9ab6-287470919e0c.jpg" width="256">

6. **Execute the code**: Open `python` and execute the following code snippet: 
```python
from pupyl.search import PupylImageSearch

# data_dir can be any created (if not, pupyl will create for you),
# or even you can leave it blank to save the assets to your
# temporary directory
SEARCH = PupylImageSearch(data_dir='/path/to/store/assets')

# Index a source (this case, a directory) that contains duplicates.
SEARCH.index('/path/with/image/duplicates', check_unique=True)
```
7. **If everything works as intended, the following error must be raised**:
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pupyl/pupyl/search.py", line 170, in index
    index.append(
  File "pupyl/pupyl/indexer/facets.py", line 187, in append
    raise UniqueItemError
pupyl.indexer.exceptions.UniqueItemError
```

Any further information, please not hesitate to ask.

**Reply by policratus** [reply]
@92amartins, just opened #69 fixing #66.

**Reply by 92amartins** [reply]
Your instructions are very straightforward, so maybe pupyl don't need the clutter of releasing alpha and beta versions right now.

I was able to reproduce the indexing process and raised an error when the `check_unique` flag was set to true 🎉 

Thanks for the effort :)
